### PR TITLE
Fix LightningCLI saving hyperparameters breaking change

### LINKED ISSUE #19977: With yaml config file for LightningCLI, `self.save_hyperparameters()` behavior abnormal
### Bug description

With yaml config file for LightningCLI, `self.save_hyperparameters()` in `__init__` of the model and datamodule mistakenly saves a `dict` containing keys like `class_path` and `init_args`.

This problems appears in version 2.3.0, but version 2.2.5 works correctly.

### What version are you seeing the problem on?

2.3.0

### How to reproduce the bug

config.yaml
```yaml
ckpt_path: null
seed_everything: 0
model:
  class_path: model.Model
  init_args:
    learning_rate: 1e-2
data:
  class_path: datamodule.DataModule
  init_args:
    data_dir: data
trainer:
  accelerator: auto
  strategy: auto
  devices: auto
  num_nodes: 1
  precision: null
  fast_dev_run: false
  max_epochs: 100
  min_epochs: null
  max_steps: -1
  min_steps: null
  max_time: null
  limit_train_batches: null
  limit_val_batches: 10
  limit_test_batches: null
  limit_predict_batches: null
  logger:
    class_path: lightning.pytorch.loggers.TensorBoardLogger
    init_args:
      save_dir: lightning_logs
      name: normalized
  callbacks:
    class_path: lightning.pytorch.callbacks.ModelCheckpoint
    init_args:
      save_top_k: 5
      monitor: valid_loss
      filename: "{epoch}-{step}-{valid_loss:.8f}"
  overfit_batches: 0.0
  val_check_interval: 50
  check_val_every_n_epoch: 1
  num_sanity_val_steps: null
  log_every_n_steps: 50
  enable_checkpointing: null
  enable_progress_bar: null
  enable_model_summary: null
  accumulate_grad_batches: 1
  gradient_clip_val: null
  gradient_clip_algorithm: null
  deterministic: true
  benchmark: null
  inference_mode: true
  use_distributed_sampler: true
  profiler: null
  detect_anomaly: false
  barebones: false
  plugins: null
  sync_batchnorm: true
  reload_dataloaders_every_n_epochs: 0
  default_root_dir: null
```

model.py
```python
import torch
from torch import nn
import torch.nn.functional as F
import lightning as pl


class Model(pl.LightningModule):
    def __init__(self, learning_rate: float):
        super().__init__()

        print()
        print("Model:")

        print(f"learning_rate: {learning_rate}")
        ## This outputs correctly.

        self.save_hyperparameters()
        
        print(self.hparams)
        ## This outputs:
        # "_instantiator": lightning.pytorch.cli.instantiate_module
        # "class_path":    model.Model
        # "init_args":     {'learning_rate': 0.01}

```

datamodule.py
```python
from lightning import LightningDataModule
from torch.utils.data import DataLoader

from dataset import KaptchaDataset
from transform import Transform


class DataModule(LightningDataModule):
    def __init__(self, data_dir: str):
        super().__init__()
        self.save_hyperparameters()

        print()
        print("DataModule:")

        print(self.hparams)
        ## This outputs
        # "_instantiator": lightning.pytorch.cli.instantiate_module
        # "class_path":    datamodule.DataModule
        # "init_args":     {'data_dir': 'data'}
```

main.py
```python
from lightning.pytorch.cli import LightningCLI
from lightning.pytorch.loggers import TensorBoardLogger
from lightning.pytorch.callbacks import ModelCheckpoint

from model import Model
from datamodule import DataModule


def cli_main():
    cli = LightningCLI()


if __name__ == "__main__":
    cli_main()
```

Run `python main.py fit --config config.yaml`
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### PR DESCRIPTION
**Pull request by mauvilsa**
## What does this PR do?

Changes the `LightningCLI`'s `load_from_checkpoint` support such that the hparams are not saved with the `class_path` and `init_args` structure, since this constitutes a breaking change which we failed to notice in #18105.

Fixes #19977

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20068.org.readthedocs.build/en/20068/

### CODE REVIEW

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Thanks for the quick fix @mauvilsa. ~This seems very specific to `init_args` and ignores `dict_kwargs`.~ I also seem to be missing many class parameters with default values that are not listed in the YAML file, and/or optional values that default to None.

EDIT: It's possible no special handling is needed for `dict_kwargs`.

**Comment by adamjstewart** [comment]
Steps to reproduce are identical to https://github.com/Lightning-AI/pytorch-lightning/issues/19977#issuecomment-2214014470, but replace `model.py` with:
```python
from lightning.pytorch import LightningModule


class Model(LightningModule):
    def __init__(self, learning_rate, step_size=None):
        super().__init__()
        self.save_hyperparameters()
        assert 'learning_rate' in self.hparams
        assert 'step_size' in self.hparams

    def training_step(*args):
        pass

    def configure_optimizers(*args):
        pass
```
The code will fail because `step_size` is not defined in `self.hparams`.

**Comment by mauvilsa** [comment]
> I also seem to be missing many class parameters with default values that are not listed in the YAML file, and/or optional values that default to None.

Indeed, there was a bug.

> It's possible no special handling is needed for dict_kwargs.

It is needed. It is now fixed and I added a new unit test based on your code snippet which includes `dict_kwargs`.

@adamjstewart please check again.

**Review by adamjstewart — APPROVED** [review]
Passes all TorchGeo tests now, thanks for the quick fix!

**Review by awaelchli — APPROVED** [review]
Thank you @mauvilsa, very appreciated!

File: src/lightning/pytorch/cli.py [hunk 1]
```diff
@@ -791,17 +791,27 @@ def __init__(self, cli: LightningCLI, key: str) -> None:
         self.key = key
 
     def __call__(self, class_type: Type[ModuleType], *args: Any, **kwargs: Any) -> ModuleType:
+        hparams = self.cli.config_dump.get(self.key, {})
+        if "class_path" in hparams:
+            hparams = {
+                "_class_path": hparams["class_path"],
```
[the remark this answers is not in the document]
Comment (mauvilsa) [reply src/lightning/pytorch/cli.py:800]: Added a comment.

**Comment by awaelchli** [comment]
Thanks @mauvilsa @adamjstewart for the help here 🚀

**Pull request merged**